imagetest/testworkflow.go

```diff
@@ -0,0 +1,245 @@
+		for stepname, step := range ts.wf.Steps {
+			if !strings.HasPrefix(stepname, "wait-") {
+				continue
+			}
+			if err := ts.wf.AddDependency(copyStep, step); err != nil {
+				return fmt.Errorf("failed to add copy-objects step dependency to workflow %s: %v", ts.Name, err)
+			}
+		}
+
+	}
+	return nil
+}
+
+type testResult struct {
+	testWorkflow    *TestWorkflow
+	skipped         bool
+	workflowSuccess bool
+	err             error
+	results         []string
+}
+
+func getTestResults(ctx context.Context, ts *TestWorkflow) ([]string, error) {
+	results := []string{}
+	createVMsStep := ts.wf.Steps[createVMsStepName]
+	for _, vm := range createVMsStep.CreateInstances.Instances {
+		log.Printf("going to download %s", vm.Metadata["_test_results_url"])
+		out, err := utils.DownloadGCSObject(ctx, client, vm.Metadata["_test_results_url"])
+		if err != nil {
+			return nil, fmt.Errorf("failed to get results for test %s vm %s: %v", ts.Name, vm.Name, err)
+		}
+		results = append(results, string(out))
+	}
+
+	return results, nil
+}
+
+// NewTestWorkflow returns a new TestWorkflow.
+func NewTestWorkflow(name, image string) (*TestWorkflow, error) {
+	t := &TestWorkflow{}
+	t.Name = name
+	t.Image = image
+
+	parts := strings.Split(image, "/")
+	t.ShortImage = parts[len(parts)-1]
+
+	t.wf = daisy.New()
+	t.wf.Name = strings.ReplaceAll(name, "_", "-")
+
+	return t, nil
+}
+
+// PrintTests prints all test workflows.
+func PrintTests(ctx context.Context, testWorkflows []*TestWorkflow, project, zone string) {
+	finalizeWorkflows(testWorkflows, zone, project)
+	for _, test := range testWorkflows {
+		if test.wf == nil {
+			continue
+		}
+		test.wf.Print(ctx)
+	}
+}
+
+// ValidateTests validates all test workflows.
+func ValidateTests(ctx context.Context, testWorkflows []*TestWorkflow, project, zone string) error {
+	finalizeWorkflows(testWorkflows, zone, project)
+	for _, test := range testWorkflows {
+		if test.wf == nil {
+			continue
+		}
+		if err := test.wf.Validate(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RunTests runs all test workflows.
+func RunTests(ctx context.Context, testWorkflows []*TestWorkflow, outPath, project, zone string, parallelCount int) {
+	var err error
+	client, err = storage.NewClient(ctx)
+	if err != nil {
+		log.Fatalf("Failed to set up storage client: %v", err)
+	}
+	finalizeWorkflows(testWorkflows, zone, project)
+
+	testResults := make(chan testResult, len(testWorkflows))
+	testchan := make(chan *TestWorkflow, len(testWorkflows))
+
+	var wg sync.WaitGroup
+	for i := 0; i < parallelCount; i++ {
+		wg.Add(1)
+		go func(id int) {
+			defer wg.Done()
+			for test := range testchan {
+				testResults <- runTestWorkflow(ctx, test)
+			}
+		}(i)
+	}
+	for _, ts := range testWorkflows {
+		testchan <- ts
+	}
+	close(testchan)
+	wg.Wait()
+
+	var suites testSuites
+	for i := 0; i < len(testWorkflows); i++ {
+		suites.TestSuite = append(suites.TestSuite, parseResult(<-testResults))
```

Comment:
Curious, is `<-testResult` even reached, since wg.Wait() will wait until test runners are done. Test runners output result to `testResults` but there's no one to read from them. Maybe I'm misunderstanding the logic.

Reply:
we create N goroutines, give them a test channel and a results channel. each goroutine is using `range` to read from testchan, placing results in results channel, and exit when it is empty
then we load the testchan with all workflows to run, and immediately close the channel
we wait for all goroutines to finish
then we pull all of the results out of the result channel serially and parse them

ime this is a fairly standard form of goroutine based parallelization, sometimes called worker goroutines